utils/src/test/java/software/amazon/awssdk/utils/cache/lru/LruCacheTest.java

```diff
@@ -0,0 +1,372 @@
+        }
+
+        /**
+         * Wait for a certain number of "gets" to have finished. This will time out and fail the test after a certain amount of
+         * time if the "gets" never finish.
+         */
+        public void waitForGetsToHaveFinished(int numExpectedGets) {
+            assertTrue(invokeSafely(() -> finishedGetPermits.tryAcquire(numExpectedGets, 10, TimeUnit.SECONDS)));
+        }
+
+        /**
+         * Release all threads blocked in this supplier.
+         */
+        @Override
+        public void close() {
+            permits.release(50);
+        }
+    }
+
+
+    private ExecutorService executorService;
+    private List<Future<?>> allExecutions;
+    private Supplier<LruCache<Key, Value>> simpleCache = () -> LruCache.builder(simpleValueSupplier)
+                                                                       .maxSize(MAX_SIMPLE_CACHE_SIZE)
+                                                                       .build();
+
+
+
+    @BeforeEach
+    public void setup() {
+        executorService = Executors.newFixedThreadPool(50);
+        allExecutions = new ArrayList<>();
+    }
+
+
+    @AfterEach
+    public void shutdown() {
+        executorService.shutdown();
+    }
+
+    @Test
+    void when_cacheHasMiss_ValueIsCalculatedAndCached() {
+        LruCache<Key, Value> cache = simpleCache.get();
+        populateAndVerify(cache, 1);
+    }
+
+    @Test
+    void when_cacheHasHit_ValueIsRetrievedFromCache() {
+        LruCache<Key, Value> cache = simpleCache.get();
+
+        populateAndVerify(cache, MAX_SIMPLE_CACHE_SIZE);
+        Value result = cache.get(simpleTestKeys.get(2));
+        assertThat(cache.size()).isEqualTo(MAX_SIMPLE_CACHE_SIZE);
+        assertThat(result).isNotNull();
+        assertThat(result).isEqualTo(simpleTestValues.get(2));
+
+        cache.get(simpleTestKeys.get(2));
+
+        int onlyOneInvocation = 1;
+        verify(simpleValueSupplier, times(onlyOneInvocation)).apply(simpleTestKeys.get(2));
+    }
+
+    @Test
+    void when_cacheFillsUp_ValuesAreEvictedFromCache() {
+        LruCache<Key, Value> cache = simpleCache.get();
+
+        populateAndVerify(cache, MAX_SIMPLE_CACHE_SIZE);
+
+        //evict 1
+        Value result = cache.get(simpleTestKeys.get(4));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 4);
+
+        //move 3 up
+        result = cache.get(simpleTestKeys.get(3));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 3);
+
+        //evict 2
+        result = cache.get(simpleTestKeys.get(5));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 5);
+
+        //move 4 up
+        result = cache.get(simpleTestKeys.get(4));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 4);
+
+        //evict 3
+        result = cache.get(simpleTestKeys.get(2));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 2);
```

Comment:
All of our evictions are in the original ordering (1, 2, 3), so it doesn't seem like we're actually testing that we're "moving things up". Should we evict 3 earlier than 2 so that we know the 'moving up' is working?

Reply:
This test case was refactored and improved.